File: examples/pod-workload-secret/go/astis_workload_test.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"strings"
	"testing"
)

// Cross-language known-answer test: the fixture is sealed by the verified Python
// reference implementation; this Go client opens it — proving both ends agree on the
// ASTIS:v2 wire contract (the same contract CVS uses). If suite IDs, the empty-AAD
// convention, the info string, or the binding AAD drift, this fails.

func loadFixture(t *testing.T) map[string]string {
	t.Helper()
	raw, err := os.ReadFile("kat-fixture.json")
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f map[string]string
	if err := json.Unmarshal(raw, &f); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	return f
}

func TestGoOpensPythonSealedEnvelope(t *testing.T) {
	f := loadFixture(t)
	env, err := parseEnvelope(f["envelope_wire"])
	if err != nil {
		t.Fatal(err)
	}
	info := []byte("astis-rfc020|" + env.RequestID + "|" + env.KekRef)
	dek, err := hpkeOpen(b64d(f["rewrap_enc"]), b64d(f["rewrap_hpke_ct"]), b64d(f["recip_priv"]), info)
	if err != nil {
		t.Fatalf("hpke open: %v", err)
	}
	if len(dek) != 32 {
		t.Fatalf("DEK len = %d, want 32", len(dek))
	}
	pt, err := aesGCMDecrypt(env, dek)
	if err != nil {
		t.Fatalf("aes-gcm: %v", err)
	}
	if string(pt) != f["expect"] {
		t.Fatalf("plaintext = %q, want %q", pt, f["expect"])
	}
}

func TestWrongBindingAADFails(t *testing.T) {
	f := loadFixture(t)
	env, _ := parseEnvelope(f["envelope_wire"])
	info := []byte("astis-rfc020|" + env.RequestID + "|" + env.KekRef)
	dek, _ := hpkeOpen(b64d(f["rewrap_enc"]), b64d(f["rewrap_hpke_ct"]), b64d(f["recip_priv"]), info)
	env.WorkloadBinding["namespace"] = "attacker-ns"
	if _, err := aesGCMDecrypt(env, dek); err == nil {
		t.Fatal("decrypt must fail when the binding AAD is tampered")
	}
}

func TestHPKEOpenTamperFails(t *testing.T) {
	f := loadFixture(t)
	env, _ := parseEnvelope(f["envelope_wire"])
	info := []byte("astis-rfc020|" + env.RequestID + "|" + env.KekRef)
	enc := b64d(f["rewrap_enc"])
	enc[0] ^= 0x01
	if _, err := hpkeOpen(enc, b64d(f["rewrap_hpke_ct"]), b64d(f["recip_priv"]), info); err == nil {
		t.Fatal("hpke open must fail on a corrupted encapsulation")
	}
}

func TestRejectsBadPrefix(t *testing.T) {
	if _, err := parseEnvelope("ASTIS:v1:whatever"); err == nil {
		t.Fatal("must reject a non-v2 prefix")
	}
}

func TestEnvelopeFieldsDecode(t *testing.T) {
	f := loadFixture(t)
	wire := strings.TrimPrefix(f["envelope_wire"], "ASTIS:v2:")
	if _, err := base64.RawURLEncoding.DecodeString(strings.TrimRight(wire, "=")); err != nil {
		t.Fatalf("envelope body not base64url: %v", err)
	}
}

File: examples/pod-workload-secret/nodejs/astis-workload.test.js
// Cross-language known-answer test (node --test). The fixture is sealed by the verified
// Python reference; this Node client opens it — proving both ends agree on the ASTIS:v2
// wire contract (the same contract CVS uses).

import test from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import { hpkeOpen, aesGcmDecrypt, parseEnvelope } from "./astis-workload.js";

const fixture = JSON.parse(fs.readFileSync(new URL("./kat-fixture.json", import.meta.url)));
const b64 = (s) => Buffer.from(s, "base64");

function recoverDek(f, env) {
  const info = Buffer.from(`astis-rfc020|${env.request_id}|${env.kek_ref}`);
  return hpkeOpen(b64(f.rewrap_enc), b64(f.rewrap_hpke_ct), b64(f.recip_priv), info);
}

test("node opens python-sealed envelope", () => {
  const env = parseEnvelope(fixture.envelope_wire);
  const dek = recoverDek(fixture, env);
  assert.equal(dek.length, 32);
  assert.equal(aesGcmDecrypt(env, dek).toString(), fixture.expect);
});

test("wrong binding AAD fails", () => {
  const env = parseEnvelope(fixture.envelope_wire);
  const dek = recoverDek(fixture, env);
  env.workload_binding.namespace = "attacker-ns";
  assert.throws(() => aesGcmDecrypt(env, dek));
});

test("hpke open tamper fails", () => {
  const env = parseEnvelope(fixture.envelope_wire);
  const info = Buffer.from(`astis-rfc020|${env.request_id}|${env.kek_ref}`);
  const enc = b64(fixture.rewrap_enc);
  enc[0] ^= 0x01;
  assert.throws(() => hpkeOpen(enc, b64(fixture.rewrap_hpke_ct), b64(fixture.recip_priv), info));
});

test("rejects non-v2 prefix", () => {
  assert.throws(() => parseEnvelope("ASTIS:v1:whatever"));
});
